crates/bevy_core_widgets/src/callback.rs

```diff
@@ -111,3 +113,100 @@ impl Notify for DeferredWorld<'_> {
         }
     }
 }
+
+/// A component that hangs on to a registered one-shot system, and unregisters it when the
+/// component is despawned.
+#[derive(Component)]
+#[component(on_remove = on_despawn_callback_owner::<I>, storage = "SparseSet")]
+pub struct OwnedCallbackSystem<I: SystemInput + Send>(SystemId<I, ()>);
+
+fn on_despawn_callback_owner<I: SystemInput + Send + 'static>(
+    mut world: DeferredWorld,
+    context: HookContext,
+) {
+    let system_id = world
+        .entity(context.entity)
+        .get::<OwnedCallbackSystem<I>>()
+        .unwrap()
+        .0;
+    world.commands().unregister_system(system_id);
+}
+
+/// Methods for registering scoped callbacks.
+pub trait RegisterOwnedCallback {
+    /// Registers a scoped one-shot system, with no input, that will be removed when the parent
+    /// entity is despawned.
+    fn register_owned_callback<M, I: IntoSystem<(), (), M> + 'static>(
+        &mut self,
+        callback: I,
+    ) -> Callback;
+
+    /// Registers a scoped one-shot systemm, with input, that will be removed when the
+    /// parent entity is despawned.
+    fn register_owned_callback_with<
+        M,
+        A: SystemInput + Send + 'static,
+        I: IntoSystem<A, (), M> + 'static,
+    >(
+        &mut self,
+        callback: I,
+    ) -> Callback<A>;
+}
+
+impl RegisterOwnedCallback for EntityCommands<'_> {
+    fn register_owned_callback<M, I: IntoSystem<(), (), M> + 'static>(
+        &mut self,
+        callback: I,
+    ) -> Callback {
+        let system_id = self.commands().register_system(callback);
+        let owner = self.id();
+        self.commands()
+            .spawn((OwnedCallbackSystem(system_id), crate::owner::OwnedBy(owner)));
```

Comment:
Registered systems are entities themselves, so you might be able to skip having an extra entity by doing something like: 

```suggestion
        self.commands()
            .entity(system_id.entity())
            .insert(crate::owner::OwnedBy(owner));
```

I haven't thought through all the consequences of that, but thought it might be worth mentioning!